django/db/backends/postgresql/compiler.py

```diff
@@ -0,0 +1,49 @@
+from django.db.models.sql.compiler import (
+    SQLAggregateCompiler,
+    SQLCompiler,
+    SQLDeleteCompiler,
+)
+from django.db.models.sql.compiler import SQLInsertCompiler as BaseSQLInsertCompiler
+from django.db.models.sql.compiler import SQLUpdateCompiler
+
+__all__ = [
+    "SQLAggregateCompiler",
+    "SQLCompiler",
+    "SQLDeleteCompiler",
+    "SQLInsertCompiler",
+    "SQLUpdateCompiler",
+]
+
+
+class InsertUnnest(list):
+    """
+    Sentinel value to signal `DatabaseOperations.bulk_insert_sql` that the
+    UNNEST strategy should be used for the bulk insert.
+    """
+
+    def __str__(self):
+        return "unnest(%s)" % ", ".join(self)
+
+
+class SQLInsertCompiler(BaseSQLInsertCompiler):
+    def assemble_as_sql(self, fields, value_rows):
+        # Specialize bulk-insertion of literal non-array values through
+        # unnest to reduce the time spent planning the query.
+        if (
+            len(value_rows) <= 1
+            # Raw fields or compilable cannot be combined in an array
+            # of literal values.
+            or any(field is None for field in fields)
+            or any(
+                any((hasattr(value, "as_sql") for value in row)) for row in value_rows
+            )
+        ):
+            return super().assemble_as_sql(fields, value_rows)
+        db_types = [field.db_type(self.connection) for field in fields]
+        # The unnest strategy cannot be used if any of the fields is an array
+        # because it cannot discriminate on the depth of unnesting.
+        if any(db_type.endswith("[]") for db_type in db_types):
+            return super().assemble_as_sql(fields, value_rows)
+        return InsertUnnest(["(%%s)::%s[]" % db_type for db_type in db_types]), [
+            list(map(list, zip(*value_rows)))
+        ]
```

Comment:
Hi Simon, I just happened upon that unnest blog post in my feed as well then saw this, nice 👍 

I don't suppose we can get some blank lines in here to help readability? 😅